keps/sig-cli/20190805-pod-troubleshooting.md

```diff
@@ -0,0 +1,380 @@
+   arguments. This will include an optional container for namespace targeting
+   (described below).
+2. Fetch the specified pod's existing ephemeral containers using
+   `GetEphemeralContainers()` in the generated pod client.
+3. Append the new debug container to the pod's ephemeral containers and call
+   `UpdateEphemeralContainers()`.
+4. [optional] Watch pod for updates to the debug container's `ContainerStatus`
+   and automatically attach once the container is running.
+
+### Debug Container Naming
+
+Currently, there is no support for deleting or recreating ephemeral containers.
+In cases where the user does not specify a name, `kubectl` will generate a
+unique name and display it to the user.
+
+### Container Namespace Targeting
+
+In order to see processes running in other containers in the pod, [Process
+Namespace Sharing] should be enabled for the pod. In cases where process
+namespace sharing isn't enabled for the pod, `kubectl` will set
+`TargetContainer` in the `EphemeralContainer`. This will cause the ephemeral
+container to be created in the namespaces of the target container in runtimes
+that support container namespace targeting.
+
+[Process Namespace Sharing]: https://kubernetes.io/docs/tasks/configure-pod-container/share-process-namespace
+
+### Interactive Troubleshooting and Automatic Attaching
+
+Since the primary use case for `kubectl debug` is interactive troubleshooting,
+`kubectl debug` will automatically attach to the console of the newly created
+ephemeral container and will default to creating containers with `Stdin` and
+`TTY` enabled.
+
+These may be disabled via command line flags.
+
+### Proposed `kubectl debug` arguments
+
+```
+% kubectl help debug
+Execute a container in a pod.
+
+Examples:
+  # Get network address configuration from pod mypod
+  kubectl debug mypod ip addr list
+
+  # Start an interactive debugging session with a debian image
+  kubectl debug mypod --image=debian
+
+Options:
+  -a, --attach=true: Automatically attach to container once created.
+  -c, --container='': Container name. If omitted one will be chosen.
+  -i, --stdin=true: Pass stdin to the container
+  -m, --image='busybox': Container image to use for debug container.
+  -p, --target='': Target processes in this container name.
+  -t, --tty=true: Stdin is a TTY
+
+Usage:
+  kubectl debug (POD | TYPE/NAME) [-c CONTAINER] [flags] -- COMMAND [args...] [options]
+
+Use "kubectl options" for a list of global command-line options (applies to all commands).
+```
+
+### User Stories
+
+#### Operations
+
+Alice runs a service "neato" that consists of a statically compiled Go binary
+running in a minimal container image. One of the its pods is suddenly having
+trouble connecting to an internal service. Being in operations, Alice wants to
+be able to inspect the running pod without restarting it, but she doesn't
+necessarily need to enter the container itself. She wants to:
```

Comment:
That argument falls short, I'd prefer to use the same argument you used at the beginning, that she doesn't have the ability to enter the container b/c it's a scratch image.

Reply:
The scenario I'm envisioning is attaching a debugger to capture state of a hard-to-reproduce bug when it occurs in production. Restarting the container will destroy the state.